### DISCUSSION TITLE
Add additional block start/stop methods in addition to init/deinit?

### DISCUSSION THREAD

**Discussion by etschneider** [FutureSDR/FutureSDR#86] [Ideas]
Radio source/sink blocks make a good example.

Currently the device and stream are initialized in init(). However some devices can take non-negligible amount time to initialize (so one radio block might start streaming seconds before another). Additionally, it is likely preferable to not start streaming at all until all blocks have been successfully initialized.

Init/deinit will control the device initialization and cleanup, while start/stop will control the stream.

The runtime would step though init(), start(), stop(), deinit() as the flowgraph life-cycle.

This would be irrelevant for most blocks, but blocks with external connections (or heavy initialization overhead) would likely benefit. Additionally, this adds the ability to start/stop a flowgraph without requiring device re-initialization.

**Comment by bastibl** [comment]
We definitely need a better idea what the protocol and guarantees with `init` and `deinit` are (e.g., in which thread context they are executed and what happens when the flowgraph is paused/stopped, reconfigured, and restarted). So at some point, we might need such methods. But before introducing them, I think we should have a better idea about flowgraph reconfiguration and, therefore, the final life-cycle of a flowgraph. If we add start/stop now, we still don't solve the problem. We only make it more unlikely or less severe, since there is still no synchronization between sources.

Also, it looks like, we are already waiting for all blocks to `init()`, before triggering the first `work()` iteration:
https://github.com/FutureSDR/FutureSDR/blob/main/src/runtime/runtime.rs#L233-L262

But the problem you describe is still there, since streaming is initiated in `init()`. It would, however, be possible to modify the block to start streaming in the first call to work. That means, it would not strictly be necessary introduce start/stop to achieve what you describe.

**Reply by etschneider** [reply]
I agree about the need for more discussion regarding lifecycle, reconfiguration, and such. I would be glad to participate if you started a new discussion on the topic.

Re: radio sync, I'm simply using radios as an instance of a more general case, something that takes a long time to initialize, regardless of sample alignment. For example one source could overflow its buffer by the time the second source finishes initializing.

As you state, the start issue could be handled in the first work, but how then the stop case? A stop message presumably. That leads to potential issues with the message queue timing and differences in the block implementations of `work()`. IMO, it is better to have obvious block kernel trait methods rather that rely on block implementers to know how to do the right thing. I totally respect the desire to not act prematurely though.

I do need to sped more time in the scheduler and messaging logic so I can comment more intelligently...

**Reply by bastibl** [reply]
What I thought about recently (not sure if it's a good idea) is to call `init` and `deinit` exactly once during the lifetime and `start`/`stop` every time the flowgraph is paused or reconfigured. Let's say we have a `Head` block: It is, generally, not clear whether it should forward the configured number of samples only once or whether the counter should be reset after a flowgraph reconfiguration. With these life-cycle methods, it would be trivial to differentiate: Reset the counter in `init` -> only do it once; reset the counter in `start` -> forward samples again after reconfiguration.

Still need a better idea about how flowgraph reconfiguration should work, though.

**Reply by etschneider** [reply]
That is what I would expect the life-cycle to be as well.

Reconfiguration could certainly be tricky, and I doubt there are any perfect solutions with respect to buffer and message queue states.

An even more tricky (potential) feature: partial reconfiguration while running. E.g. Keep a radio interface running while stopping and reconfiguring some other processing blocks. There is a lot to go wonky there, but I figured I'd put the idea out there for you to think about. I'm not really sure it would be worth the trouble, but it would be a cool trick.
